### PR TITLE
*: add check when watching

### PR DESCRIPTION
**Pull request by zimulala**
When we call `Watch` we need to check if it works normally. And the watch channel may be closed for some reasons. So we need to rewatch it when it is closed.

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -338,11 +339,18 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
 			err := do.Reload()
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
+			if !ok {
+				log.Info("[ddl] reload schema in loop, schema syncer need rewatch")
+				// Make sure the rewatch doesn't affect load schema, so we let the timeout is equal to lease/2.
+				ctx, cancelFunc := context.WithTimeout(context.Background(), interval)
+				syncer.WatchGlobalSchemaVer(ctx, 1)
+				cancelFunc()
```
Comment (shenli) [inline domain/domain.go]: Why cancel here?

File: ddl/syncer.go [hunk 2]
```diff
@@ -175,6 +180,32 @@ func (s *schemaVersionSyncer) GlobalVersionCh() clientv3.WatchChan {
...
+		var err error
+		select {
+		case resp, ok := <-ch:
+			if !ok || resp.Err() != nil {
+				err = resp.Err()
+				break
+			}
+			// This situation should not happen.
+			s.globalVerCh = ch
+			log.Warnf("[syncer] watch global schema finished, but this situation should not happen")
```
Comment (shenli) [inline ddl/syncer.go]: If this is an unexpected situation, why not return an error?

File: ddl/syncer.go [hunk 3]
```diff
@@ -175,6 +180,32 @@ func (s *schemaVersionSyncer) GlobalVersionCh() clientv3.WatchChan {
...
+			return nil
+		default:
+			s.globalVerCh = ch
+			log.Info("[syncer] watch global schema finished")
+			return nil
+		}
+		log.Warnf("[syncer] watch global schema failed %v", err)
+		time.Sleep(keyOpRetryInterval)
```
Comment (shenli) [inline ddl/syncer.go]: The `keyOpRetryInterval` is 30ms. If pd is down, there will be a lot of logs?

File: domain/domain.go [hunk 4]
```diff
@@ -338,11 +339,18 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
+		case _, ok := <-syncer.GlobalVersionCh():
 			err := do.Reload()
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
+			if !ok {
+				log.Info("[ddl] reload schema in loop, schema syncer need rewatch")
+				// Make sure the rewatch doesn't affect load schema, so we let the timeout is equal to lease/2.
+				ctx, cancelFunc := context.WithTimeout(context.Background(), interval)
+				syncer.WatchGlobalSchemaVer(ctx, 1)
```
Comment (shenli) [inline domain/domain.go]: The return value is ignored.
Do we need a metrics here?

File: domain/domain.go [hunk 5]
```diff
@@ -338,11 +339,20 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
+		case _, ok := <-syncer.GlobalVersionCh():
 			err := do.Reload()
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
+			if !ok {
+				log.Info("[ddl] reload schema in loop, schema syncer need rewatch")
+				// Make sure the rewatch doesn't affect load schema, so we let the timeout is equal to lease/2.
+				ctx, _ := context.WithTimeout(context.Background(), interval)
+				err = syncer.WatchGlobalSchemaVer(ctx, 1)
```
Comment (winkyao) [inline domain/domain.go]: Can we use a constant to replace hard-code number `1`?

File: ddl/syncer.go [hunk 6]
```diff
@@ -175,6 +180,39 @@ func (s *schemaVersionSyncer) GlobalVersionCh() clientv3.WatchChan {
...
+		ch := s.etcdCli.Watch(ctx, DDLGlobalSchemaVersion)
+		var err error
+		select {
+		case resp, ok := <-ch:
+			if !ok || resp.Err() != nil {
+				err = resp.Err()
+				break
+			}
+			s.globalVerCh = ch
+			log.Warnf("[syncer] watch global schema finished, and it has a resp %#v", resp)
```
Comment (winkyao) [inline ddl/syncer.go]: Should we return the resp? Let outside can do a reload?

File: domain/domain.go [hunk 7]
```diff
@@ -338,11 +339,20 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
-		case <-syncer.GlobalVersionCh():
+		case _, ok := <-syncer.GlobalVersionCh():
 			err := do.Reload()
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
+			if !ok {
+				log.Info("[ddl] reload schema in loop, schema syncer need rewatch")
```
Comment (winkyao) [inline domain/domain.go]: Add a metrics here?

File: ddl/syncer.go [hunk 8]
```diff
@@ -135,7 +137,10 @@ func (s *schemaVersionSyncer) Init(ctx context.Context) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
-	s.globalVerCh = s.etcdCli.Watch(ctx, DDLGlobalSchemaVersion)
+	err = s.WatchGlobalSchemaVer(ctx, keyOpDefaultRetryCnt)
```
Comment (coocood) [inline ddl/syncer.go]: I think we don't need to retry in `Init`.

File: domain/domain.go (hunk 5 above)
Comment (coocood) [inline domain/domain.go]: Can we simply call `ch := s.etcdCli.Watch(ctx, DDLGlobalSchemaVersion)` here?

File: domain/domain.go [hunk 9]
```diff
@@ -338,11 +339,20 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
+				// Make sure the rewatch doesn't affect load schema, so we let the timeout is equal to lease/2.
+				ctx, _ := context.WithTimeout(context.Background(), interval)
```
Comment (coocood) [inline domain/domain.go]: Will the channel close when `interval` time passed?

File: domain/domain.go (hunk 5 above)
Comment (tiancaiamao) [inline domain/domain.go]: When the timeout signal fired, would the return chan by Watch been closed?

File: domain/domain.go [hunk 10]
```diff
@@ -338,11 +339,20 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop err %v", errors.ErrorStack(err))
 			}
-		case <-syncer.GlobalVersionCh():
+		case _, ok := <-syncer.GlobalVersionCh():
```
Comment (tiancaiamao) [inline domain/domain.go]: How about combine `WatchGlobalSchemaVer` and `GlobalVersionCh` into one function?

```
select {
    case <- f():
}
```

`f()` will be called every time the select statement runs.

**Comment by shenli** [comment]
Please fix the CI.

**Comment by coocood** [comment]
LGTM

File: domain/domain.go [hunk 11]
```diff
@@ -338,11 +338,16 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
```
Comment (tiancaiamao) [inline domain/domain.go]: `<-syncer.GlobalVersionCh()` is likely to be `nil` here.

**Comment by tiancaiamao** [comment]
LGTM

File: domain/domain.go (hunk 11 above)
Comment (zimulala) [reply domain/domain.go]: Yes, it may be `nil` here. Then This case doesn't be entered.

**Pull request merged**